doc/specs/#5000 - Process Model 2.0/#5000 - Process Model 2.0.md

```diff
@@ -0,0 +1,1253 @@
+extension of this scenario being "run a `wt` commandline in _any_ given WT
+window".
+
+This spec by no means attempts to fully document how this functionality should
+work. This scenarios deserves it's own spec to discuss various different naming
+
+schemes for the commandline parameters. The following is given as an example of
+how these arguments _might_ be authored and implemented to satisfy some of these
+scenarios.
+
+Since each window process will have it's own unique ID assigned to it by the
+monarch, then running a command in a given window with ID `N` should be as easy
+as something like:
+
+```sh
+wt.exe --session N new-tab ; split-pane
+```
+
+(or for shorthand, `wt -s N new-tab ; split-pane`).
+
+This would create a new peasant, who could then ask the monarch if there is a
+peasant with ID `N`. If there is, then the peasant can connect to that other
+peasant, dispatch the current commandline to that other peasant, and exit,
+before ever creating a window. If this commandline instead creates a new monarch
+process, then there was _no_ other monarch, and so there must logically not be
+any other existing WT windows, and the `--session N` argument could be safely
+ignored, and the command run in the current window.
+
+We should reserve the session id `0` to always refer to "The current window", if
+there is one. So `wt -s 0 new-tab` will run `new-tab` in the current window (if
+we're being run from WT).
+
+If `wt -s 0 <commands>` is run _outside_ a WT instance, it could attempt to glom
+onto _the most recent WT window_ instead. This seems more logical than something
+like `wt --session last` or some other special value indicating "run this in the
+MRU window".
+
+That might be a simple, but **wrong**, implementation for "the current window".
+If the peasants always raise an event when their window is focused, and the
+monarch keeps track of the MRU order for peasants, then one could naively assume
+that the execution of `wt -s 0 <commands>` would always return the window the
+user was typing in, the current one. However, if someone were to do something
+like `sleep 10 ; wt -s 0 <commands>`, then the user could easily focus another
+WT window during the sleep, which would cause the MRU window to not be the same
+as the window executing the command.
+
+I'm not sure that there is a better solution for the `-s 0` scenario other than
+attempting to use the `WT_SESSION` environment variable. If a `wt.exe` process is
+spawned and that's in it's environment variables, it could try and ask the
+monarch for the peasant who's hosting the session corresponding to that GUID.
+This is more of a theoretical solution than anything else.
+
+In Single-Instance mode, running `wt -s 0` outside a WT window will still cause
+the commandline to glom to the existing single terminal instance, if there is
+one.
+
+#### Scenario: Quake Mode
+
+"Quake mode" has a variety of different scenarios<sup>[[1]](#footnote-1)</sup>

```

Comment:
Based on my limited experience with Linux DEs, terminals that support (and enabled) "Quake mode" are likely to allow only one instance. If you try to open another instance, it will just activate the terminal in quake style. The global hotkey only need to do one thing, that is try to run the terminal program. Maybe this way the quake mode implementation can be simplified?

Reply:
I I doubt that people who want "Quake mode" would also need multiple WT windows. What's the point, right? If enabling "Quake Mode" just enables "Single Instance Mode" mandatorily , I think people should be OK with it. And the implementation could be simplified dramatically.

Reply:
For the record, I'm intentionally trying to be vague about quake mode details here - I'd rather give that feature the full analysis of a dedicated spec. I believe it deserved a callout in this doc mostly due to the possible overlap with single instance mode, and how a monarch might make registering the hotkey easier.

Once we can register a hotkey in a resilient manner, then we can do whatever we want with it ☺️